### PR TITLE
add perm_chain_assignment randomization

### PR DESCRIPTION
**Pull request by snuffysasa**
I plan to keep working on this some more and features related to it.

But right now it seems to be working and tests are passing, so up to you if you want to work on merging into main now, or wait.  Doesn't matter much to me for this PR.

### CODE REVIEW

File: src/randomizer.py [hunk 1]
```diff
@@ -1936,6 +1936,62 @@ def visit_Assignment(self, node: ca.Assignment) -> None:
...
+                    or ast_util.equal_ast(stmt.rvalue, next_stmt.rvalue)
+                )
+                and not ast_util.is_effectful(next_stmt.rvalue)
+            ):
+                cands.append((i, block))
+
+            ast_util.for_nested_blocks(stmt, rec)
+
+        # call on last statement in block, because it was skipped above
+        ast_util.for_nested_blocks(statements[-1], rec)
```
Comment (simonlindholm) [inline src/randomizer.py:1972]:
```suggestion
        for stmt in statements:
            ast_util.for_nested_blocks(stmt, rec)
```

File: src/randomizer.py [hunk 2]
```diff
@@ -1936,6 +1936,62 @@ def visit_Assignment(self, node: ca.Assignment) -> None:
...
+        statements = ast_util.get_block_stmts(block, False)
+        numStatements = len(statements)
+
+        if numStatements < 2:
+            return
+
+        # don't include the last statement in list
+        for i in range(numStatements - 1):
+            stmt = statements[i]
+            next_stmt = statements[i + 1]
```
[the remark this answers is not in the document]
Comment (snuffysasa) [reply src/randomizer.py:1989]: wow! looks like black magic to me!!

[the remark this answers is not in the document]
Comment (snuffysasa) [reply src/randomizer.py:1989]: but cool, I'll update with this

**Comment by simonlindholm** [comment]
Can you move the toml file changes to another PR?

**Comment by snuffysasa** [comment]
> Can you move the toml file changes to another PR?

Yea that makes sense.

**Comment by snuffysasa** [comment]
okay now this PR does not include the weights in a toml file

File: src/randomizer.py [hunk 3]
```diff
@@ -1936,6 +1936,52 @@ def visit_Assignment(self, node: ca.Assignment) -> None:
     ast_util.insert_statement(tob, toi, dup)
 
 
+def perm_chain_assignment(
+    fn: ca.FuncDef, ast: ca.FileAST, indices: Indices, region: Region, random: Random
+) -> None:
+    """Combine two consecutive assignments into one chain assignment."""
+    cands: List[Tuple(int, ca.Compound)] = []
```
Comment (simonlindholm) [inline src/randomizer.py:1943]:
```suggestion
    cands: List[Tuple[int, ca.Compound]] = []
```

File: src/randomizer.py [hunk 4]
```diff
@@ -1936,6 +1936,52 @@ def visit_Assignment(self, node: ca.Assignment) -> None:
...
+    rec(fn.body)
+
+    ensure(cands)
+    chosen_assignment_idx, block = random.choice(cands)
+
+    statements = ast_util.get_block_stmts(block, False)
+    stmt = statements[chosen_assignment_idx]
+    next_stmt = statements[chosen_assignment_idx + 1]
+
+    ## Insert the new lvalue into left or right side of old lvalue
```
Comment (simonlindholm) [inline src/randomizer.py:1973]:
```suggestion
    # Insert the new lvalue into left or right side of old lvalue
```

File: src/randomizer.py [hunk 5]
```diff
@@ -1936,6 +1936,52 @@ def visit_Assignment(self, node: ca.Assignment) -> None:
...
+
+        for stmt in statements:
+            ast_util.for_nested_blocks(stmt, rec)
+
+    rec(fn.body)
+
+    ensure(cands)
+    chosen_assignment_idx, block = random.choice(cands)
+
+    statements = ast_util.get_block_stmts(block, False)
```
Comment (simonlindholm) [inline src/randomizer.py:1969]:
```suggestion
    statements = ast_util.get_block_stmts(block, True)
```
(doesn't matter but it's semantically right)

**Review by simonlindholm — APPROVED** [review]

**Pull request merged**